### ISSUE TITLE
Dynamic attrs in factories (functions) are called twice 

### ISSUE THREAD

**Issue by cibernox** [samselikoff/ember-cli-mirage#910]
This arose while doing this:

```js
Factory.extend({
  id() {
    return autoincrementalIdSharedBySeveralModels();
  }
});
```

This code generates ids `[2, 4, 6, 8, 10]` etc...

I dig a bit into why this happens and it's because mirage allows functions to be called out of order. By example, the next code is allowed:

```
  let BazFactory = Mirage.Factory.extend({
    bar() {
      return this.foo + 2; // This function calls foo as a property despite of being a function
    },

    baz: 6,

    foo(i) {
      return this.baz * i;
    }
  });
```
To allow this mirage has to evaluate the functions twice, once to generate a Directed Acyclic Graph to know the exection order and the second to actually call them in order.

It's not easy to fix, but I think that if there functions were defined using `getters/setters`, none of this would be needed. However this outcome seems so obvious and the current solution so unintuitive that It makes me think that there is a reason I don't see why things are the way they are.

Was it because of IE8 support? Other?

**Comment by samselikoff** [comment]
Someone else implemented the graph code here and I haven't looked at it in a while. The change was made to allow dynamic factory attrs, like in your example (`bar` depends on `foo` so `foo` needs to be calculated first). It's basically a lightweight computed property system.

When you say they're called twice you're saying the array should be `[1, 2, 3, 4]`? Also I'm not clear what you mean by using getters - are you referring to the code internal to Mirage's factory?

**Comment by cibernox** [comment]
Yes, the generated sequence should be `[1,2,3,4]`, but since for every `server.create` invocation the functions are called twice (the first time the result is discarded, it's just to figure out the execution order), the generated values are `[2, 4, 6, 8]`.

By using getters I mean that functions are stored as getters in the factory.
By user a getters, there is no need of swiping the attribute definitions twice, one to create the dependency graph and another to get the actual values. Since getter functions can access another getters, the dependency graph is implicit. Circular dependencies wouldn't be supported, but neither are now.

**Comment by samselikoff** [comment]
I see. 

Guarav just made an update to the factory - I wonder if it addressed the issue? And if not, I could use some help to re-work the system the way you describe. Maybe when you (or anyone else!) has some time?

**Comment by cibernox** [comment]
I'm in the middle of a secret project that I'd like to release this week in ember-fest, so I don't foreseen having enough time in the next 2 weeks, but I'll try to get back to this, maybe during office ours.

**Comment by samselikoff** [comment]
I verified the behavior and agree it's confusing. I think at this point the path forward of changing the API to use getters instead of functions is a big enough change that it should be considered alongside other plans to change Mirage classes over to vanilla ES6 classes.

For now I'm going to mark this as a feature request, and we can come back to it when we move the rest of the objects over to ES6.
